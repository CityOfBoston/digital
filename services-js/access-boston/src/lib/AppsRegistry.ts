import yaml from 'js-yaml';

export interface AppsCategory {
  title: string;
  showRequestAccessLink: boolean;
  icons: boolean;
  apps: App[];
}

export interface App {
  title: string;
  url: string;
  iconUrl: string | null;
  description: string;
  // null groups means "everyone can see this"
  groups: string[] | null;
  mfaDeviceRequired: boolean;
  // null agencies means "all agencies"
  agencies: string[] | null;
  target: string;
}

/**
 * This class is in lib rather than server just so we can use it in Storybook
 * stories. It doesnâ€™t actually get used by the client app.
 */
export default class AppsRegistry {
  showAll: boolean;
  allCategories: AppsCategory[];

  constructor(appsYaml: any, showAll = false) {
    this.showAll = showAll;
    const yamlCategories = appsYaml.categories;

    if (!yamlCategories || !Array.isArray(yamlCategories)) {
      throw new Error('Missing categories array');
    }

    this.allCategories = yamlCategories.map(c => {
      const { title, apps: yamlApps, show_request_access_link, icons } = c;

      if (!title || typeof title !== 'string') {
        throw new Error('Category missing title: ' + JSON.stringify(c));
      }

      if (!yamlApps || !Array.isArray(yamlApps)) {
        throw new Error('Category missing apps array: ' + JSON.stringify(c));
      }

      const apps: App[] = yamlApps.map(a => {
        const {
          title,
          url,
          groups,
          description,
          icon,
          mfa_device_required,
          agencies,
          target,
        } = a;

        if (!title || typeof title !== 'string') {
          throw new Error('App missing a title: ' + JSON.stringify(a));
        }

        if (!url || typeof url !== 'string') {
          throw new Error('App missing a url: ' + JSON.stringify(a));
        }

        if (groups && !Array.isArray(groups)) {
          throw new Error('groups is not an array: ' + JSON.stringify(a));
        }

        if (agencies && !Array.isArray(agencies)) {
          throw new Error('agencies is not an array: ' + JSON.stringify(a));
        }

        return {
          title,
          url,
          iconUrl: icon || null,
          description: description || '',
          groups: groups || null,
          mfaDeviceRequired: mfa_device_required || false,
          agencies: agencies || null,
          target: target || '',
        };
      });

      return {
        title,
        apps,
        showRequestAccessLink: !!show_request_access_link,
        icons: !!icons,
      };
    });
  }

  appsForGroups(
    userGroups: string[],
    hasMfaDevice: boolean,
    cobAgency: string | null
  ): AppsCategory[] {
    return (
      this.allCategories
        .map(c => ({
          ...c,
          apps: c.apps.filter(({ groups, mfaDeviceRequired, agencies }) => {
            const groupsRequirementMet =
              !groups || groups.find(g => userGroups.includes(g));

            const mfaRequirementMet = !mfaDeviceRequired || hasMfaDevice;

            const agencyRequirementMet =
              !agencies || (cobAgency && agencies.includes(cobAgency));

            return (
              this.showAll ||
              (groupsRequirementMet &&
                mfaRequirementMet &&
                agencyRequirementMet)
            );
          }),
        }))
        // Filter out apps with no categories
        .filter(c => c.apps.length > 0)
    );
  }
}

export function makeAppsRegistry(
  yamlString: string,
  showAll = false
): AppsRegistry {
  const appsYaml = yaml.safeLoad(yamlString);
  return new AppsRegistry(appsYaml, showAll);
}
