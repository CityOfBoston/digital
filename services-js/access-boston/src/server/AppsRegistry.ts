import fs from 'fs';
import yaml from 'js-yaml';

export interface AppsCategory {
  title: string;
  showRequestAccessLink: boolean;
  icons: boolean;
  apps: App[];
}

export interface App {
  title: string;
  url: string;
  iconUrl?: string;
  description: string;
  // null groups means "everyone can see this"
  groups: string[] | null;
}

export default class AppsRegistry {
  showAll: boolean;
  allCategories: AppsCategory[];

  constructor(appsYaml: any, showAll = false) {
    this.showAll = showAll;
    const yamlCategories = appsYaml.categories;

    if (!yamlCategories || !Array.isArray(yamlCategories)) {
      throw new Error('Missing categories array');
    }

    this.allCategories = yamlCategories.map(c => {
      const { title, apps: yamlApps, show_request_access_link, icons } = c;

      if (!title || typeof title !== 'string') {
        throw new Error('Category missing title: ' + JSON.stringify(c));
      }

      if (!yamlApps || !Array.isArray(yamlApps)) {
        throw new Error('Category missing apps array: ' + JSON.stringify(c));
      }

      const apps: App[] = yamlApps.map(a => {
        const { title, url, groups, description, icon } = a;

        if (!title || typeof title !== 'string') {
          throw new Error('App missing a title: ' + JSON.stringify(a));
        }

        if (!url || typeof url !== 'string') {
          throw new Error('App missing a url: ' + JSON.stringify(a));
        }

        if (groups && !Array.isArray(groups)) {
          throw new Error('groups is not an array: ' + JSON.stringify(a));
        }

        return {
          title,
          url,
          iconUrl: icon,
          description: description || '',
          groups: groups || null,
        };
      });

      return {
        title,
        apps,
        showRequestAccessLink: !!show_request_access_link,
        icons: !!icons,
      };
    });
  }

  appsForGroups(userGroups: string[]): AppsCategory[] {
    return (
      this.allCategories
        .map(c => ({
          ...c,
          apps: c.apps.filter(
            ({ groups }) =>
              this.showAll ||
              !groups ||
              !!groups.find(g => userGroups.includes(g))
          ),
        }))
        // Filter out apps with no categories
        .filter(c => c.apps.length > 0)
    );
  }
}

export async function makeAppsRegistry(yamlPath: string, showAll = false) {
  const yamlString = await new Promise<string>((resolve, reject) => {
    fs.readFile(yamlPath, 'utf-8', (err, contents: string) => {
      if (err) {
        reject(err);
      } else {
        resolve(contents);
      }
    });
  });

  const appsYaml = yaml.safeLoad(yamlString, { filename: yamlPath });
  return new AppsRegistry(appsYaml, showAll);
}
