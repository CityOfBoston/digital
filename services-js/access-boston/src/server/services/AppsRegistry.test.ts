import fs from 'fs';
import path from 'path';
import yaml from 'js-yaml';

import AppsRegistry from './AppsRegistry';

describe('appsForUserTypeAndGroups', () => {
  let appsRegistry: AppsRegistry;

  beforeEach(() => {
    const yamlFixture = yaml.safeLoad(
      fs.readFileSync(
        path.resolve(__dirname, '../../../fixtures/apps.yaml'),
        'utf-8'
      )
    );

    appsRegistry = new AppsRegistry(yamlFixture);
  });

  it('returns just the "everyone" category for no groups', () => {
    expect(appsRegistry.appsForGroups([])).toMatchSnapshot();
  });

  it('returns apps for a group', () => {
    // This checks that categories with no apps are not returned
    expect(appsRegistry.appsForGroups(['SG_AB_IAM_TEAM'])).toMatchSnapshot();
  });
});
