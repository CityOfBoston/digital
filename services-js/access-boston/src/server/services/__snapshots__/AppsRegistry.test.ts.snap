// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appsForUserTypeAndGroups returns apps for a group 1`] = `
Array [
  Object {
    "apps": Array [
      Object {
        "description": "",
        "groups": null,
        "iconUrl": undefined,
        "title": "The Hub",
        "url": "https://hub.boston.gov/",
      },
      Object {
        "description": "",
        "groups": null,
        "iconUrl": "/static/apps/boston-maps.png",
        "title": "Boston Maps",
        "url": "https://boston.maps.arcgis.com/",
      },
    ],
    "icons": true,
    "showRequestAccessLink": true,
    "title": "Applications",
  },
  Object {
    "apps": Array [
      Object {
        "description": "",
        "groups": null,
        "iconUrl": undefined,
        "title": "Building Maintenance",
        "url": "https://eamtst.cityhall.boston.cob:5443/WebMaint/",
      },
      Object {
        "description": "",
        "groups": null,
        "iconUrl": undefined,
        "title": "FleetHub",
        "url": "http://boston.getlocalmotion.com/",
      },
      Object {
        "description": "",
        "groups": null,
        "iconUrl": undefined,
        "title": "International Travel",
        "url": "https://goo.gl/forms/Xobh9aZeGXSnKH2m2",
      },
      Object {
        "description": "",
        "groups": null,
        "iconUrl": undefined,
        "title": "Unblock Website Request",
        "url": "https://boston.seamlessdocs.com/f/geol18cumhw",
      },
    ],
    "icons": false,
    "showRequestAccessLink": false,
    "title": "Forms",
  },
  Object {
    "apps": Array [
      Object {
        "description": "",
        "groups": null,
        "iconUrl": undefined,
        "title": "Change Password",
        "url": "/change-password",
      },
      Object {
        "description": "",
        "groups": null,
        "iconUrl": undefined,
        "title": "Change MFA Device or Email Address",
        "url": "https://identity.boston.gov/identityiq/changePassword.jsf",
      },
    ],
    "icons": false,
    "showRequestAccessLink": false,
    "title": "Account Tools",
  },
]
`;

exports[`appsForUserTypeAndGroups returns just the "everyone" category for no groups 1`] = `
Array [
  Object {
    "apps": Array [
      Object {
        "description": "",
        "groups": null,
        "iconUrl": undefined,
        "title": "The Hub",
        "url": "https://hub.boston.gov/",
      },
      Object {
        "description": "",
        "groups": null,
        "iconUrl": "/static/apps/boston-maps.png",
        "title": "Boston Maps",
        "url": "https://boston.maps.arcgis.com/",
      },
    ],
    "icons": true,
    "showRequestAccessLink": true,
    "title": "Applications",
  },
  Object {
    "apps": Array [
      Object {
        "description": "",
        "groups": null,
        "iconUrl": undefined,
        "title": "Building Maintenance",
        "url": "https://eamtst.cityhall.boston.cob:5443/WebMaint/",
      },
      Object {
        "description": "",
        "groups": null,
        "iconUrl": undefined,
        "title": "FleetHub",
        "url": "http://boston.getlocalmotion.com/",
      },
      Object {
        "description": "",
        "groups": null,
        "iconUrl": undefined,
        "title": "International Travel",
        "url": "https://goo.gl/forms/Xobh9aZeGXSnKH2m2",
      },
      Object {
        "description": "",
        "groups": null,
        "iconUrl": undefined,
        "title": "Unblock Website Request",
        "url": "https://boston.seamlessdocs.com/f/geol18cumhw",
      },
    ],
    "icons": false,
    "showRequestAccessLink": false,
    "title": "Forms",
  },
  Object {
    "apps": Array [
      Object {
        "description": "",
        "groups": null,
        "iconUrl": undefined,
        "title": "Change Password",
        "url": "/change-password",
      },
      Object {
        "description": "",
        "groups": null,
        "iconUrl": undefined,
        "title": "Change MFA Device or Email Address",
        "url": "https://identity.boston.gov/identityiq/changePassword.jsf",
      },
    ],
    "icons": false,
    "showRequestAccessLink": false,
    "title": "Account Tools",
  },
]
`;
